nselib/msrpctypes.lua

```diff
@@ -4522,6 +4522,28 @@ function unmarshall_SERVICE_STATUS(data, pos)
   return pos, result
 end
 
+--- Unmarshalls a null-terminated Unicode string based upon a 32-bit offset (LPTSTR)
+-- @param arguments The data being processed
+-- @param pos  The current position within the data
+-- @return The new position
+-- @return The string with null removed
+function unmarshall_lptstr(arguments, startpos)
+
+  local offset = 5
```

Comment:
Why is the offset 5? This function should handle reading the offset, extracting the string, and returning the new position and the string. In pseudocode:

```
function unmarshall_lptstr(args, pos)
  pos, offset = unmarshall_int32(args, pos)
  endpos = offset
  while endpos < #args
    if args:sub(endpos, endpos+2) == "\0\0"
      break
    end
  end
  return pos, args:sub(offset, endpos)
```

Reply:
Offset 5 must be added to line [4545](https://github.com/nmap/nmap/pull/987/files#diff-34f3f003b9e7a1620fbd3a0bbaccdeb5R4545) because while receive the response from the server the first 4 bytes are allocated for displaying the size of the packet and so we have to start unmarshalling from the 5th position.

While we extract the starting position of the strings from the binary response, we get the starting position wrt to 1, since we were starting unmarshalling from 5th position, 5 has to be added to the unpack function.

If you think this will be a problem, we can change this by passing a parameter as offset like

```
function unmarshall_lptstr(arguments, startpos, offset)
offset = offset or 0
......
end
```

Does this look nice?

Reply:
Your idea of using `string.sub` degrades the performance because when we know the end position there is no point in iterating all chars finding for NULL byte. I think its better to pass endpos argument to `unmarshall_lptstr` to improve the performance.

Your views?